packages/webui/src/client/ui/RundownView/SelectedElementsContext.tsx

```diff
@@ -0,0 +1,165 @@
+import React from 'react'
+import {
+	AdLibActionId,
+	PartInstanceId,
+	PieceId,
+	RundownId,
+	SegmentId,
+} from '@sofie-automation/corelib/dist/dataModel/Ids'
+
+interface RundownElement {
+	type: 'rundown'
+	id: string
+	elementId: RundownId
+}
+
+interface SegmentElement {
+	type: 'segment'
+	id: string
+	elementId: SegmentId
+}
+
+interface PartInstanceElement {
+	type: 'partInstance'
+	id: string
+	elementId: PartInstanceId
+}
+
+interface PieceElement {
+	type: 'piece'
+	id: string
+	elementId: PieceId
+}
+
+interface AdlibActionElement {
+	type: 'adlibAction'
+	id: string
+	elementId: AdLibActionId
+}
+
+// Union type for all possible elements
+type SelectedElement = RundownElement | SegmentElement | PartInstanceElement | PieceElement | AdlibActionElement
+
+interface SelectionContextType {
+	selectedElements: Map<string, SelectedElement>
+	isSelected: (elementId: string) => boolean
+	clearAndSetSelection: (element: SelectedElement) => void
+	toggleSelection: (element: SelectedElement) => void
+	addSelection: (element: SelectedElement) => void
+	removeSelection: (elementId: string) => void
+	clearSelections: () => void
+	getSelectedCount: () => number
+}
+
+type SelectionAction =
+	| { type: 'CLEAR_AND_SET_SELECTION'; payload: SelectedElement }
+	| { type: 'TOGGLE_SELECTION'; payload: SelectedElement }
+	| { type: 'ADD_SELECTION'; payload: SelectedElement }
+	| { type: 'REMOVE_SELECTION'; payload: string }
+	| { type: 'CLEAR_SELECTIONS' }
+
+// Reducer function
+const selectionReducer = (
+	state: Map<string, SelectedElement>,
+	action: SelectionAction,
+	maxSelections: number
+): Map<string, SelectedElement> => {
+	switch (action.type) {
+		case 'CLEAR_AND_SET_SELECTION': {
+			const newMap = new Map([[action.payload.id, action.payload]])
+			return newMap
+		}
+		case 'TOGGLE_SELECTION': {
+			const next = new Map(state)
+			if (next.has(action.payload.id)) {
+				next.delete(action.payload.id)
+			} else if (next.size < maxSelections) {
+				next.set(action.payload.id, action.payload)
+			}
+			return next
+		}
+		case 'ADD_SELECTION': {
+			if (state.size >= maxSelections) return state
```

Comment:
This feels weird. This means that if `maxSelections` is 1, then I will not be able to select anything until I deselect something. That doesn't sit right, I think, UX-wise.

I would propose that if `state.size >= maxSelections` then the very first item from the map is removed (capitalizing on the fact that keys are ordered in a Map in the order of insertion) and a new key is allowed to be set.

Reply:
in the initial draft the maxSelection was a constant, maybe it's not necessary to have. Currently it's set default to 10. But maybe we should set it higher. But a maxSelections of 1 would be sort of using the code wrong, as ADD_SELECTION indicates that it is possible to add more than the currently selected.
Maybe that number just should be 10000. 
But not having a max could maybe cause some extreme amounts of selections, which was why it's there.

Reply:
Okay, I still don't think this is a good idea. Whatever `maxSelections` is, with this setup, it's impossible to select something, until I deselect something, if `maxSelections` is reached. I guess I can see that `ADD_SELECTION` means add up until `maxSelection` and then `TOGGLE_SELECTION` is generally what one would trigger on a click, but the problem with `maxSelections` running out still stands.